File: src/commands/group/list.ts
import {Command, flags as flagtype} from "@oclif/command";
import chalk from "chalk";
import {ironnode} from "../../lib/SDK";
import {createDisplayTable} from "../../lib/Utils";
import {keyFile} from "../../lib/sharedFlags";

/**
 * Group list command. Retrieves all groups that the user is a part of and displays them in a table to the console.
 */
export default class List extends Command {
    static description = "Display a list of all the groups of which you're either an admin or member.";
    static flags = {
        help: flagtype.help({char: "h"}),
        keyfile: keyFile(),
    };

    async run() {
        this.parse(List);
        ironnode()
            .group.list()
            .then((groups) => {
                if (groups.result.length === 0) {
                    return this.log("You aren't currently an admin or member of any groups.");
                }
                const table = createDisplayTable(["Group Name", "Group ID", "Admin", "Member"]);

                const check = chalk.green("✔");
                const nope = chalk.red("✖");
                groups.result.map((group) => {
                    table.push([group.groupName, group.groupID, group.isAdmin ? check : nope, group.isMember ? check : nope]);
                });
                this.log(`\n${table.toString()}\n`);
            })
            .catch((e) => {
                this.log(chalk.red(e.message));
                this.error(chalk.red("Failed to retrieve list of groups."));
            });
    }
}

File: src/group/list.rs
use crate::util;
use clap::Parser;
use ironoxide::prelude::BlockingIronOxide;
use prettytable::{color, Attr, Cell, Row};
use std::path::PathBuf;

#[derive(Parser)]
/// Display a list of all the groups of which you're either an admin or member.
pub struct List {
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(parse(from_os_str), short, long)]
    keyfile: Option<PathBuf>,
}

impl util::GetKeyfile for List {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

pub fn list_groups(sdk: &BlockingIronOxide) -> Result<(), String> {
    let groups = sdk.group_list().unwrap();

    if groups.result().is_empty() {
        Err("You aren't currently an admin or member of any groups.".to_string())
    } else {
        let mut table =
            table!([Fbb=>"Group Name", "Admin", "Member", "Group ID", "Created", "Updated"]);
        for group in groups.result() {
            let check = Cell::new("✓").with_style(Attr::ForegroundColor(color::GREEN));
            let nope = Cell::new("✗").with_style(Attr::ForegroundColor(color::RED));
            let is_admin = if group.is_admin() {
                check.clone()
            } else {
                nope.clone()
            };
            let is_member = if group.is_member() {
                check.clone()
            } else {
                nope.clone()
            };
            table.add_row(Row::new(vec![
                cell!(Fw -> group.name().map(|n| n.name().as_str()).unwrap_or("")),
                is_admin,
                is_member,
                cell!(Fw -> group.id().id()),
                cell!(Fw -> group.created()),
                cell!(Fw -> group.last_updated()),
            ]));
        }

        table.printstd();

        Ok(())
    }
}
